### PR TITLE
Support parsing of checked user-defined operator declarations

### PR DESCRIPTION
**Pull request by AlekseyTs**
https://github.com/dotnet/csharplang/blob/main/proposals/checked-user-defined-operators.md

Language version checks and whether `checked` is applicable to specific operator will be checked during declaration binding and is out of scope for this PR.

Relates to test plan https://github.com/dotnet/roslyn/issues/59458

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -3305,8 +3305,23 @@ private ConversionOperatorDeclarationSyntax TryParseConversionOperatorDeclaratio
 
                 if (style.IsMissing)
                 {
-                    if (this.CurrentToken.Kind != SyntaxKind.OperatorKeyword || SyntaxFacts.IsAnyOverloadableOperator(this.PeekToken(1).Kind) ||
+                    bool possibleConversion = true;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:3308]: nit: remove initial value.  it is never observed.

**Comment by CyrusNajmabadi** [comment]
NIt: consider detecting, reporting, and gracefully recovering from errant use of `unchecked` in the location where `checked` is used.  I can see people easily trying that out and it would potentially be good to have the experience gracefully handle that.

**Review by CyrusNajmabadi — APPROVED** [review]
Grammar/Parser changes LGTM.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs [hunk 2]
```diff
@@ -2514,6 +2514,33 @@ public static BlockSyntax Block(IEnumerable<StatementSyntax> statements)
                 expressionBody: expressionBody);
         }
 
+        /// <summary>Creates a new ConversionOperatorDeclarationSyntax instance.</summary>
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs:2517]: nit: `<see cref="..." />` (applies to below methods as well)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@333fred, @chsienki, @dotnet/roslyn-compiler Please review.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1). You could address Cyrus's comment about unnecessary local initialization in LanguageParser in follow-up PR

**Review by 333fred — APPROVED** [review]
LGTM (commit 1)

**Pull request merged**